### PR TITLE
Tutorial

### PR DESCRIPTION
**Pull request by deniznida**

### CODE REVIEW

**Comment by AnnJohn** [comment]
Assigned to @jmburges

File: TUTORIAL.md [hunk 1]
```diff
@@ -0,0 +1,392 @@
...
+
+```ruby
+def self.drop_table
+  sql = "DROP TABLE IF EXISTS courses"
+  DB[:conn].execute(sql)
+end
+```
+Run `rspec` and your will see the next error, `undefined method insert`. Again, lets take a look at our `student.rb` file and see if we have reusable code. 
+
+It looks like we have an `insert` method built for us so lets modify that and use it. Let's also create a helper method called `attribute_values` to keep everything organized. Instead of typing out all of the values in our insert method, we can call the `attribute_values` method instead.
```
Comment (jmburges) [inline TUTORIAL.md]: Wait how did you know that the next thing to work on was the insert method? What's the goal behind modifiying it?

File: TUTORIAL.md [hunk 2]
```diff
@@ -0,0 +1,392 @@
...
+    (name, department_id)
+    VALUES
+    (?,?)
+  SQL
+  DB[:conn].execute(sql, attribute_values)
+
+  @id = DB[:conn].execute("SELECT last_insert_rowid() FROM courses")[0][0]
+end
+```
+Now when we run `rspec`, we see `undefined method new_from_db`. This method is going to take the information from the database and create a new instance by utilizing a method called `tap`. The `tap` method "yields `self` to the block, and then returns `self`. The primary purpose of this method is to “tap into” a method chain, in order to perform operations on intermediate results within the chain." In our case we are tapping into Course and creating a new instance while assigning it's attributes to specific rows.
```
Comment (jmburges) [inline TUTORIAL.md]: How did you know "This method is going to take the information from the database and create a new instance"? 

Can you compare `tap` to something else? I find it hard to understand for many people. Like what would you have to write if the `tap` method didn't exist?

File: TUTORIAL.md [hunk 3]
```diff
@@ -0,0 +1,392 @@
...
+def self.new_from_db(row)
+  self.new.tap do |s|
+    s.id = row[0]
+    s.name =  row[1]
+    s.department_id = row[2]
+  end
+end
+```
+
+Now we can create a Courses table, insert a course and drop a course. Being able to find a course by it's name would also be useful, which is our next error.
```
Comment (jmburges) [inline TUTORIAL.md]: Maybe show how to prove this in `binding.pry`? We really gotta show them how to test things without tests!

File: TUTORIAL.md [hunk 4]
```diff
@@ -0,0 +1,392 @@
...
+    s.id = row[0]
+    s.name =  row[1]
+    s.department_id = row[2]
+  end
+end
+```
+
+Now we can create a Courses table, insert a course and drop a course. Being able to find a course by it's name would also be useful, which is our next error. 
+
+We need to build a method called `find_by_name`. Again we can use `lib/student.rb` file as an example, replacing the necessary variables. Here we are selecting all from our courses table where the name is equal to the param we are passing in, then limiting it to one.
```
Comment (jmburges) [inline TUTORIAL.md]: how did oyu know that?

File: TUTORIAL.md (hunk 1 above)
Comment (deniznida) [inline TUTORIAL.md]: @jmburges I believe @danny-sullivan created this tutorial I just opened the pr into solution

File: TUTORIAL.md [hunk 5]
```diff
@@ -0,0 +1,392 @@
...
+      WHERE name = ?
+      LIMIT 1
+    SQL
+    result = DB[:conn].execute(sql,name)
+    result.map do |row| 
+      self.new_from_db(row)
+    end.first
+  end
+```
+Let's run `rspec` for our next error, `undefined method find_all_by_department_id`. Here we want to be able to find all courses by their `department_id`. The `SELECT` statement for this reads pretty clearly, `SELECT *` (`*` stands for all) from the courses table, where the `deparment_id` of the course equals the `id` we are passing in to the method, very simlilar to finding by name. Compare this code with your Student class to see the differences.
```
Comment (jmburges) [inline TUTORIAL.md]: So for this one. I get that you knew the next thing to work on was `find_all_by_department_id` but how did you know that it wants "to find all courses by their department_id" was it just the method name? the test?

Comment (jmburges) [inline TUTORIAL.md]: also `deparment_id` typo

File: TUTORIAL.md [hunk 6]
```diff
@@ -0,0 +1,392 @@
...
+  result.collect do |row| 
+    self.new_from_db(row)
+  end
+end
+```
+Now we can create a Courses table, insert a course into the table, delete the table, find a course by name or department_id. What if we want to change the title of a course? We need a way to update the table.
+
+Run `rspec` and you will see `undefined method update`. Let's take a look at our `student.rb` file and see if there is anything to guide us. Looks like there is an `update` method built already, let's deconstruct it.
+
+The `update` method will take the values you are changing, find the object by it's id and update the instance. You can see our helper method `attribute_values` being passed in to the `DB[:conn].execute(sql, attribute_values, id)` part of the method.
```
Comment (jmburges) [inline TUTORIAL.md]: Maybe *this* is the point where you refactor the attribute_values out to a separate method!

File: TUTORIAL.md [hunk 7]
```diff
@@ -0,0 +1,392 @@
...
+def update
+  sql = <<-SQL
+    UPDATE courses
+    SET name = ?,department_id = ?
+    WHERE id = ?
+  SQL
+  DB[:conn].execute(sql, attribute_values, id)
+end
+```
+The last piece of this puzzle is the ability to save our changes, which we will write two methods for that work together.
```
Comment (jmburges) [inline TUTORIAL.md]: why write two methods? Less logical leaps!

File: TUTORIAL.md [hunk 8]
```diff
@@ -0,0 +1,392 @@
...
+    UPDATE courses
+    SET name = ?,department_id = ?
+    WHERE id = ?
+  SQL
+  DB[:conn].execute(sql, attribute_values, id)
+end
+```
+The last piece of this puzzle is the ability to save our changes, which we will write two methods for that work together. 
+
+The first is `persisted?`. This is a helper method that checks whether or not there is already a row in the database for a given object. If it is persisted it returns true, if not it returns false.
```
Comment (jmburges) [inline TUTORIAL.md]: true/false should be in backticks

File: TUTORIAL.md [hunk 9]
```diff
@@ -0,0 +1,392 @@
...
+  DB[:conn].execute(sql, attribute_values, id)
+end
+```
+The last piece of this puzzle is the ability to save our changes, which we will write two methods for that work together. 
+
+The first is `persisted?`. This is a helper method that checks whether or not there is already a row in the database for a given object. If it is persisted it returns true, if not it returns false. 
+
+In ruby, it is convention that methods ending in a `?` will return either true or false. 
+
+Skip down to the solution and you will notice the double exclamation points in the body of our `persisted?` method. These are called bangs. One bang will negate a boolean. So if something is true, a bang in front will make it false. If it is false, a bang in front will make it true.
```
Comment (jmburges) [inline TUTORIAL.md]: I'm not super opinionated on this, but I hate calling excalamation points "bang". Too sexual

File: TUTORIAL.md [hunk 10]
```diff
@@ -0,0 +1,392 @@
...
+In our case `self.id` would return an integer, by calling the double bang it returns a boolean. This way in our save method, we can call `persisted?`, which is a now boolean. 
+
+We are also using the ternary operator which reads like this.
+
+`if_this_is_a_true_value ? then_the_result_is_this : else_it_is_this`
+
+If you put it all together we get the following code for the `save` method that will allow us to check if there's already a row in the database for the object it was called on, if there is it will just `update` the data, if not it will `insert` a new row for that object.
+
+```ruby
+def persisted?
```
Comment (jmburges) [inline TUTORIAL.md]: why does checking for id see if this is persisted?

**Comment by jmburges** [comment]
Hey @ipc103 

I started reviewing this, and there are a ton of logical jumps and such. I know Dan did a bunch of this, but can you put some work into this? I'm going to create an open issue to "re-open" this PR since it's technically not possible to re-open PRs

**Comment by ipc103** [comment]
will do - should i also look at the instructions? raised #35 because the instructions were unclear to students

**Pull request merged**